File: examples/python/staged_static_kinds.py
#!/usr/bin/env python3
"""Separate a treatment effect into direct and mediated parts.

The baseline variable Z affects treatment A, mediator M, and outcome Y.
For a change from 0.2 to 0.8, the simulated direct effect is 1.8 and the
indirect effect through M is 4.8. Their total is 6.6, which also equals the
mean individual effect in this simulation. Omitting Z would bias the analysis.

Install with `python -m pip install antecedent`; see examples/README.md."""

from __future__ import annotations

import numpy as np
from antecedent import Bayesian, Counterfactual, MediationEffect, analyze, prepare


def confounded_scm(n: int = 500) -> tuple[dict[str, np.ndarray], list[tuple[str, str]]]:
    i = np.arange(n, dtype=float)
    z = np.cos(i * 0.41)
    a = np.sin(i * 0.71) + 0.4 * z
    m = 2 * a + 0.5 * z + np.cos(i * 1.13)
    y = 3 * a + 4 * m + 5 * z + 0.1 * np.sin(i * 0.31)
    graph = [
        ("a", "m"),
        ("a", "y"),
        ("m", "y"),
        ("z", "a"),
        ("z", "m"),
        ("z", "y"),
    ]
    return {"a": a, "m": m, "y": y, "z": z}, graph


def main() -> None:
    data, graph = confounded_scm()
    control, active = 0.2, 0.8
    for contrast, expected in (
        ("natural_direct", 1.8),
        ("natural_indirect", 4.8),
        ("total", 6.6),
    ):
        study = prepare(
            data,
            graph=graph,
            query=MediationEffect(
                "a",
                "y",
                mediators=["m"],
                contrast=contrast,
                control_level=control,
                active_level=active,
            ),
            refute="none",
            bootstrap=0,
        )
        result = study.estimate()
        print("Calibration:", result.calibration.status)
        print(
            f"{contrast}={result.effect:.4f} estimator={result.estimate.estimator_id}"
        )
        assert abs(result.effect - expected) < 0.03, result.effect
        assert result.estimate.estimator_id == "mediation.linear"

    ite = analyze(
        data,
        graph=graph,
        query=Counterfactual("a", "y", control_level=control, active_level=active),
        refute="none",
    )
    print(
        f"mean_ite={ite.mean_ite:.4f} n={len(ite.unit_effects)} "
        f"estimator={ite.estimate.estimator_id}"
    )
    assert abs(ite.mean_ite - 6.6) < 0.03, ite.mean_ite
    assert ite.estimate.estimator_id == "gcm.fit"
    assert len(ite.unit_effects) == len(data["a"])

    bayes = analyze(
        data,
        graph=graph,
        query=MediationEffect(
            "a",
            "y",
            mediators=["m"],
            contrast="natural_direct",
            control_level=control,
            active_level=active,
        ),
        inference=Bayesian(n_draws=64),
        refute="none",
        bootstrap=0,
    )
    print(f"bayesian_nde={bayes.effect:.4f} estimator={bayes.estimate.estimator_id}")
    assert abs(bayes.effect - 1.8) < 0.2, bayes.effect
    assert bayes.posterior is not None

    bayes_cf = analyze(
        data,
        graph=graph,
        query=Counterfactual("a", "y", control_level=control, active_level=active),
        inference=Bayesian(n_draws=64),
        refute="none",
    )
    print(
        f"bayesian_ite={bayes_cf.mean_ite:.4f} estimator={bayes_cf.estimate.estimator_id}"
    )
    assert abs(bayes_cf.mean_ite - 6.6) < 0.25, bayes_cf.mean_ite
    assert bayes_cf.posterior is not None


if __name__ == "__main__":
    main()

File: examples/rust/staged_static_kinds.rs
#![allow(
    clippy::cast_possible_truncation,
    clippy::cast_precision_loss,
    clippy::many_single_char_names,
    clippy::similar_names
)]
//! Staged `MediationEffect` and `Counterfactual` on a confounded linear SCM.
//!
//! Z confounds A, M, and Y. For control 0.2 vs active 0.8 the structural
//! contrasts are NDE=1.8, NIE=4.8, total/mean ITE=6.6.
//!
//! Run: `cargo run -p antecedent --example staged_static_kinds`

use std::sync::Arc;

use antecedent::prelude::*;
use antecedent::{CounterfactualQuery, MediationContrast, MediationQuery};

fn confounded_scm(n: usize) -> (TabularData, Dag) {
    let i = 0..n;
    let z: Vec<f64> = i.clone().map(|k| (k as f64 * 0.41).cos()).collect();
    let a: Vec<f64> = i.clone().map(|k| (k as f64 * 0.71).sin() + 0.4 * z[k]).collect();
    let m: Vec<f64> =
        i.clone().map(|k| 2.0 * a[k] + 0.5 * z[k] + (k as f64 * 1.13).cos()).collect();
    let y: Vec<f64> =
        i.map(|k| 3.0 * a[k] + 4.0 * m[k] + 5.0 * z[k] + 0.1 * (k as f64 * 0.31).sin()).collect();
    let data = TabularData::from_f64_columns([
        ("a", a.as_slice()),
        ("m", m.as_slice()),
        ("y", y.as_slice()),
        ("z", z.as_slice()),
    ])
    .expect("equal-length columns");
    let mut dag = Dag::with_variables(4);
    for (s, t) in [(0, 1), (0, 2), (1, 2), (3, 0), (3, 1), (3, 2)] {
        dag.insert_directed(DenseNodeId::from_raw(s), DenseNodeId::from_raw(t)).unwrap();
    }
    (data, dag)
}

fn main() -> Result<(), CausalError> {
    let (data, dag) = confounded_scm(500);
    let control = 0.2;
    let active = 0.8;
    let ctx = ExecutionContext::for_tests(13);
    for (label, contrast, expected) in [
        ("natural_direct", MediationContrast::NaturalDirect, 1.8),
        ("natural_indirect", MediationContrast::NaturalIndirect, 4.8),
        ("total", MediationContrast::Total, 6.6),
    ] {
        let mut query = MediationQuery::binary(
            VariableId::from_raw(0),
            VariableId::from_raw(2),
            Arc::from([VariableId::from_raw(1)]),
            contrast,
        );
        query.control = Intervention::set(query.treatment, Value::f64(control));
        query.active = Intervention::set(query.treatment, Value::f64(active));
        let result = Study::tabular(data.clone())
            .graph(dag.clone())
            .query(query)
            .refute(RefuteSuite::None)
            .bootstrap_replicates(0)
            .build()?
            .run(&ctx)?;
        println!(
            "{label}={:.4} estimator={}",
            result.estimate.ate,
            result.logical_plan.estimator.as_deref().unwrap_or("?")
        );
        assert!((result.estimate.ate - expected).abs() < 0.03, "{label}={}", result.estimate.ate);
        assert_eq!(result.logical_plan.estimator.as_deref(), Some("mediation.linear"));
    }

    let query = CounterfactualQuery::new(
        VariableId::from_raw(2),
        Arc::from([Intervention::set(VariableId::from_raw(0), Value::f64(active))]),
    )
    .with_control_level(control);
    let result = Study::tabular(data)
        .graph(dag)
        .query(query)
        .refute(RefuteSuite::None)
        .bootstrap_replicates(0)
        .build()?
        .run(&ctx)?;
    let cf = result.counterfactual.as_ref().expect("unit ITE");
    println!(
        "mean_ite={:.4} n={} estimator={}",
        cf.mean_ite,
        cf.unit_effects.len(),
        result.logical_plan.estimator.as_deref().unwrap_or("?")
    );
    assert!((cf.mean_ite - 6.6).abs() < 0.03, "mean_ite={}", cf.mean_ite);
    assert_eq!(result.logical_plan.estimator.as_deref(), Some("gcm.fit"));
    assert_eq!(cf.unit_effects.len(), 500);
    Ok(())
}
